hivemind/utils/mpfuture.py

```diff
@@ -1,147 +1,164 @@
+from __future__ import annotations
+import time
 import multiprocessing as mp
 import multiprocessing.connection
-from concurrent.futures import Future, CancelledError
-from warnings import warn
+import concurrent.futures._base as base
+
 import asyncio
+from functools import lru_cache
+from typing import Optional
+
+from hivemind.utils.threading import run_in_background
 
 
-class MPFuture(Future):
-    """ Multiprocessing version of concurrent.futures.Future, interacts between two processes via Pipe """
-    STATES = 'pending', 'running', 'cancelled', 'finished', 'exception'
-    STATE_PENDING, STATE_RUNNING, STATE_CANCELLED, STATE_FINISHED, STATE_EXCEPTION = STATES
+class MPFuture(base.Future):
```

Comment:
MPFuture now implements all public methods of concurrent.futures.Future (with functionality tests).